index.js

```diff
@@ -1 +1,151 @@
+var isArrayExists = typeof Array.isArray === 'function';
+var symbolDoesNotExist = typeof Symbol === 'undefined';
+var toStringLeftSliceLength = 8;
+var toStringRightSliceLength = -1;
+/**
+ * ### typeOf (obj)
+ *
+ * Uses `Object.prototype.toString` to determine the type of an object,
+ * normalising behaviour across engine versions & well optimised.
+ *
+ * @param {Mixed} object
+ * @return {String} object type
+ * @api public
+ */
+module.exports = function typeDetect(obj) {
+  /* ! Speed optimisation
+   * Pre:
+   *   string literal     x 3,039,035 ops/sec ±1.62% (78 runs sampled)
+   *   boolean literal    x 1,424,138 ops/sec ±4.54% (75 runs sampled)
+   *   number literal     x 1,653,153 ops/sec ±1.91% (82 runs sampled)
+   *   null               x 8,448,576 ops/sec ±1.65% (59 runs sampled)
+   *   undefined          x 9,978,660 ops/sec ±1.92% (75 runs sampled)
+   *   function           x 2,556,769 ops/sec ±1.73% (77 runs sampled)
+   * Post:
+   *   string literal     x 38,564,796 ops/sec ±1.15% (79 runs sampled)
+   *   boolean literal    x 31,148,940 ops/sec ±1.10% (79 runs sampled)
+   *   number literal     x 32,679,330 ops/sec ±1.90% (78 runs sampled)
+   *   null               x 30,573,407 ops/sec ±0.58% (82 runs sampled)
+   *   undefined          x 32,363,368 ops/sec ±1.07% (82 runs sampled)
+   *   function           x 31,296,870 ops/sec ±0.96% (83 runs sampled)
+   */
+  var typeofObj = typeof obj;
+  if (typeofObj !== 'object') {
+    return typeofObj;
+  }
+
+  /* ! Spec Conformance
+   * ES5$15.2.4.2 (ES6$19.1.3.6) - Object.prototype.toString returns "[object Null]" for null object
+   * Test: `Object.prototype.toString.call(null)``
+   *  - PhantomJS 1.x == "[object DOMWindow]" (https://github.com/ariya/phantomjs/issues/13617)
+   *  - IE <=8  == "[object Object]"
+   */
+  if (obj === null) {
+    return 'null';
+  }
+
+  /* ! Spec Conformance
+   * ES5$15.2.4.2 (ES6$19.1.3.6) - Object.prototype.toString returns "[object Undefined]" for undefined object
+   * Test: `Object.prototype.toString.call(undefined)``
+   *  - PhantomJS 1.x == "[object DOMWindow]" (https://github.com/ariya/phantomjs/issues/11722)
+   *  - IE <=8  == "[object Object]"
+   */
+  if (obj === undefined) { // eslint-disable-line no-undefined
+    return 'undefined';
+  }
+
+  /* ! Speed optimisation
+   * Pre:
+   *   array literal      x 2,888,352 ops/sec ±0.67% (82 runs sampled)
+   * Post:
+   *   array literal      x 22,479,650 ops/sec ±0.96% (81 runs sampled)
+   */
+  if (isArrayExists && Array.isArray(obj)) {
+    return 'array';
+  }
+
+  if (isPrototypeOfExists && (symbolDoesNotExist || typeof obj[Symbol.toStringTag] === 'undefined')) {
+    /* ! Speed optimisation
+    * Pre:
+    *   object from null   x 2,424,320 ops/sec ±1.67% (76 runs sampled)
+    * Post:
+    *   object from null   x 3,941,469 ops/sec ±1.89% (77 runs sampled)
+    */
+    if (Object.getPrototypeOf(obj) === null) {
```

Comment:
I saw this doesn't get tested too and I can't think of any case `Object.getPrototypeOf()` returns null, but I really don't know if there is any, so, if I'm wrong please just ignore this comment.